src/coreclr/jit/rangecheck.cpp

```diff
@@ -325,6 +325,21 @@ void RangeCheck::OptimizeRangeCheck(BasicBlock* block, Statement* stmt, GenTree*
         }
     }
 
+    if (m_pCompiler->vnStore->GetVNFunc(idxVn, &funcApp) && (funcApp.m_func == (VNFunc)GT_UMOD))
+    {
+        // We can always omit bound checks for Arr[X u% Arr.Length] pattern (unsigned MOD).
+        //
+        // if arr.Length is 0 we technically should keep the bounds check, but since the expression
+        // has to throw DividedByZeroException anyway - no special handling needed.
```

Comment:
Consider:
```
BOUNDS_CHECK
   X % a.Length
   a.Length
```
And `a.Length` is zero. `optRemoveRangeCheck` will extract `ASG` side effects only, so `DividedByZeroException` will be lost.

Note it's hard (or impossible - perhaps with forward sub on a `pop`ed span indexer?) to produce this kind of tree in practice.

This is actually a broader problem, I suppose. Since `X` can be arbitrary, all non-`ASG` side effects of it will be lost.

Reply:
Can you eloborate how exactly DivByZero is lost here during side effect extraction? I could in theory check lower bound of array.length to be constant and >0 but that defeats the main purpose of this transformation, e.g. `int Bucket(int[] buckets, uint hashcode) => buckets[hashcode % buckets.Length];` definitely doesn't need any >0 length to work

Reply:
We've talked with Egor on Discord and decided to punt this problem as pre-existing for now.

Reply:
And presumably, a proper fix needs https://github.com/dotnet/runtime/issues/84248 first to avoid massive regressions